leader/election.go

```diff
@@ -1,139 +1,204 @@
+	service = strings.Split(hostname, "-")[0]
 }
 
-func notify(isLeader bool) {
-	for _, fn := range watchers {
-		fn(isLeader)
+func Register(
+	ctx context.Context,
+	namespace string,
+	onLead func(ctx gocontext.Context),
+	onStoppedLead func(),
+	onNewLeader func(identity string),
+) error {
+	if namespace == "" {
+		namespace = podNamespace
 	}
 
-}
+	ctx = ctx.WithNamespace(namespace)
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      service,
+			Namespace: namespace,
+		},
+		Client: ctx.Kubernetes().CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: hostname,
+		},
+	}
 
-func IsLeader(ctx context.Context, leaseName string) (bool, error) {
-	leases := ctx.Kubernetes().CoordinationV1().Leases(ctx.GetNamespace())
+	electionConfig := leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: true,
+		LeaseDuration:   ctx.Properties().Duration("leader.lease.duration", 30*time.Second),
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(leadCtx gocontext.Context) {
+				updateLeaderLabel(ctx)
+				onLead(leadCtx)
+			},
+			OnStoppedLeading: onStoppedLead,
+			OnNewLeader: func(identity string) {
+				if identity == hostname {
+					return
+				}
 
-	lease, err := leases.Get(ctx, leaseName, metav1.GetOptions{})
-	if err != nil {
-		return false, err
+				onNewLeader(identity)
+			},
+		},
 	}
 
-	if *lease.Spec.HolderIdentity == identity {
-		return true, nil
+	elector, err := leaderelection.NewLeaderElector(electionConfig)
+	if err != nil {
+		return err
 	}
 
-	return false, nil
+	leaderContext, cancel := gocontext.WithCancel(ctx)
+	shutdown.AddHook(func() {
+		cancel()
+
+		// give the elector some time to release the lease
+		time.Sleep(time.Second * 2)
+	})
+
+	go elector.Run(leaderContext)
+	<-ctx.Done()
+
+	return nil
 }
 
-func createOrUpdateLease(ctx context.Context, leaseName string, attempt int) (bool, string, error) {
-	if attempt > 0 {
-		time.Sleep(time.Duration(rand.Intn(5000)) * time.Millisecond)
-	}
-	if attempt >= ctx.Properties().Int("leader.lease.attempts", 3) {
-		return false, "", fmt.Errorf("failed to acquire lease %s after %d attempts", leaseName, attempt)
+// updateLeaderLabel sets leader:true label on the current pod
+// and also removes that label from all other replicas.
+func updateLeaderLabel(ctx context.Context) {
+	backoff := retry.WithMaxRetries(3, retry.NewExponential(time.Second))
+	err := retry.Do(ctx, backoff, func(_ctx gocontext.Context) error {
+		pods, err := getAllReplicas(ctx, hostname)
+		if err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to get replicas: %w", err))
+		}
+
+		for _, pod := range pods.Items {
+			var payload string
+			if pod.Name == hostname {
+				ctx.Infof("adding leader metadata from pod: %s", pod.Name)
+				payload = `{"metadata":{"labels":{"leader":"true"}}}`
```

Comment:
i think include the service name in the label e.g. `canary-checker/leader: true`